### ISSUE TITLE
Large batch processing support

### ISSUE THREAD

**Issue by sympmarc** [SharePoint/PnP-JS-Core#668]
#### Category
[ ] Enhancement

[ ] Bug

[X] Question

#### Version

Please specify what version of the library you are using: [ none - yet ]

#### Expected / Desired Behavior / Question
Initial question from Gitter:
Wondering if PnP-JS-Core will help me with processing large batch responses. Anyone needed to do this? Use case: creating 700 list items in a go, and I'm getting some timeout errors. I'd like to be able to process those errors for retries. Does the library handle any of that for me?

Continuing...
It's around 700 items per day, loaded from a CSV-like file. The process runs in an interactive page in an AngularJS-based application, where the user picks the file they want to load (stored in a Document library). I used to load them one at a time (I originally wrote this in SharePoint 2007) and then switched to using $batch when we moved to Office 365. It's been loading just fine for months as a single batch until last week, when we starting getting timeout errors daily inside the batch.

The timeout errors look like this:
```
HTTP/1.1 500 Internal Server Error
CONTENT-TYPE: application/json;odata=nometadata;streaming=true;charset=utf-8

{"odata.error":{"code":"-2146233083, Microsoft.SharePoint.Client.ClientServiceTimeoutException","message":{"lang":"en-US","value":"The operation has timed out."}}}
--batchresponse_202f2ea9-5361-48b3-b92d-c4ead9c25a5a
Content-Type: application/http
Content-Transfer-Encoding: binary
```
The status coming back is 200, so I wasn't "seeing" fact there was any error. The timeout errors start somewhere in the 400-500 item range now, but it's variable. The elapsed wall clock time is no more than a minute or so.  Something must have changed either on the Office 365 end or something my client did - neither of which I have any expectation I will be able to discover.

I'm thinking about building the logic to parse the batch response myself, but wanted to see if the library did that already. Additionally, I'll probably break the large batch into smaller ones, and also hoped the library might have a mechanisms for that already. (There isn't any evidence of either in the docs, anyway, or in the initial replies I got in Gitter.)

No matter whether you think what I'm doing makes sense, using batching is a good idea, and I had hoped there would be something here to support it. I'm building the batches with code I modified from AC's early examples linked to [here](https://docs.microsoft.com/en-us/sharepoint/dev/sp-add-ins/make-batch-requests-with-the-rest-apis). Building the batch isn't an issue; dealing with potential failures is.

**Comment by koltyakov** [comment]
Hi Marc, thanks for posting the question!
Then dealing with batches, results and errors can be managed in separate then/catch of the requests which are executed in a batch. There will be no physical request, but the library resolves the promises from the batch result.

Here is a slightly modified example (from Gitter), with batch split to sequential requests with maximum batch size as a setting:

```typescript
import { Web, List, sp } from 'sp-pnp-js';

interface IBatchCreationData {
  item: any;
  result: any;
  error: any;
}

const processBatch = async (list: List, itemEntityType: string, data: IBatchCreationData[], index = 0): Promise<any> => {
  const batchSize = 50; // Size of a batch, should be tweaked for a package size, 25-50-a bit mo shoudl work
  if (data.length > index + 1) {
    let batch = sp.web.createBatch();
    for (let len = index + batchSize; index < len && index < data.length; index += 1) {

      let dataItem = data[index];

      let success = (function(res) { // ES5 function is on purpose to bind executioning item
        // Promise is resolved on batch is executed, no phisical request per item
        this.result = res;
      }).bind(dataItem);

      let error = (function(err) { // ES5 function is on purpose to bind executioning item
        // Promise rejected on batch is executed per item
        // if err happened (e.g. payload contains an error)
        this.error = err;
      }).bind(dataItem);

      list.items.inBatch(batch).add(dataItem.item, itemEntityType)
        .then(success)
        .catch(error);

    }
    console.log(`Processing (${index} of ${data.length})... `)
    let result = await batch.execute(); // Executing batch of `batchSize`
    // callback with process visualisation, e.i. progress bar tick
    return await processBatch(list, itemEntityType, data, index);
  } else {
    return new Promise(r => r(data));
  }
};

(async () => {

  let dummyItems: any[] = [];

  // Just a dummy data
  for (let i = 0; i < 700; i += 1) {
    dummyItems.push({
      Title: `Title ${i}`
    });
  }

  dummyItems[1].NonExistingField = 'Should cause an error in second item';

  const list = sp.web.getList(`${_spPageContextInfo.webServerRelativeUrl}/Lists/MyList`);
  const entityType = await list.getListItemEntityTypeFullName();

  let data: IBatchCreationData[] = dummyItems.map(dd => {
    return {
      item: dd,
      result: null,
      error: null
    }
  });

  return await processBatch(list, entityType, data);

})()
  .then(console.log) // array of initial objects with creation results and error if any
  .catch(console.log);
```

![image](https://user-images.githubusercontent.com/7816483/34195613-97513bb6-e56f-11e7-8c15-ea601309b229.png)

Placing such kind of logic for sequential batch splitting can be too opinionated for the library, as variations of logic can be vast (writing to different lists in a batch, processing different operations, etc.). But it can be definitely good to have such a wiki page.

@patrick-rodgers, @sympmarc any ideas can or should it really be a universal kind of logic like having batch size setting and splitting a batch to separate physical batches when batch size is overflown?

**Comment by patrick-rodgers** [comment]
Thanks @sympmarc for the questiona and @koltyakov for the great write up. I agree that we shouldn't update the library for this, batching is already complicated enough :) - but would like the put together a wiki page/example folks can use in cases like this. We do provide the ability to get a good (then) or bad (catch) response per request and the above illustrates a way to do that. Using that and a bit of context you could easily retry the failed requests, either in a fresh batch or on a one off basis.

Welcome thoughts from everyone on this as well. Ultimately would like to see us put out a "large batch guidance" article - and I can push it as well as a blog post. (Or Marc can steal the content and push it out to his much larger audience :))

**Comment by sympmarc** [comment]
Good stuff! I didn't necessarily expect the library to do all my work for me (though wouldn't that be grand??!?!), but this is very helpful.

I think the main thing I need to do is break my large batches into smaller batches, as @koltyakov recommends. My hope would be that if I do that, then the adds / updates will succeed. 

I'm curious, though. When I get timeouts in my batch, the return code is still 200. I don't *think* the code above really helps with that; I'd still need to troll through all of the responses (via the promises) to see if each individual request in the batch succeeds or not. I'll need to experiment with this approach to understand it better, but it does look promising (no pun intended!).

FYI: I'm still in the "CEWP in a 'classic' page" development mode here, and the Typescript doesn't help me much as I'm still doing JavaScript with AngularJS. Since we still can't build full page applications in "modern", I don't build in "modern", because that's the majority of what my clients need.

Something definitely changed on the back end to cause my large batches to start timing out, but I know it's virtually impossible to figure out what that could be. @patrick-rodgers, if you know of anything that might have changed in the updates just prior to 13 Dec (the first day we saw the load issues), it's probably worth figuring out what the net effect was, even if you guys just discuss it at Microsoft.

**Comment by koltyakov** [comment]
> When I get timeouts in my batch, the return code is still 200

It's better to wait for @patrick-rodgers here, he's a hero of batch API.
AFAIK, the batch request returns 200 and multipart body in which are a particular HTTP codes/responses/errors as a multipart text package.

> FYI: I'm still in the "CEWP in a 'classic' page" development mode here, and the Typescript doesn't help me much as I'm still doing JavaScript with AngularJS

I know, I saw some of you latest screencasts with `spsave` and AngularJS. We also do a lot of similar CEWPs stuff for On-Prem clients (which - CEWP + source link - is the most reliable approach for On-Prem, IMO). But completely migrated to TypeScript year and a half ago. TypeScript is a real saver.
TypeScript + CEWP + PnP JS Core stuff can be easily adopted into the project with some scaffolders, like `generator-sppp` and `sp-build-tasks`, with minimal overhead.
Some of the code can be bundled from TypeScript some from the legacy scripts which is too lazy or expensive or whatever to rewrite.

**Comment by sympmarc** [comment]
TypeScript is out when you have to work on a client machine with no admin rights. I can't install anything, which would include node.js, etc. Don't even get me started. Luckily, there's a [portable version of VS Code](https://gareth.flowers/vscode-portable/), or I'd still be editing in SharePoint Designer (which they were willing to install for me). It's not that I dislike SPD, but...

**Comment by koltyakov** [comment]
> work on a client machine with no admin rights

Yeah, this is the pain and the argument. I know such cases.
How about the portable version of toolchain as a drop in package with periodically updatable dependencies as PnP JS Core, build toolchain, etc.? ;)

**Comment by sympmarc** [comment]
> AFAIK, the batch request returns 200 and multipart body with a particular HTTP codes/responses/errors as a multipart text package.

That's correct. The overall batch comes back with a 200 even if the individual requests fail. This can be disconcerting if you don't rip apart the batch response to see what's really going on.

Where I might get a 200 back for the batch, I may still get something like this for a specific request.
```
HTTP/1.1 500 Internal Server Error
CONTENT-TYPE: application/json;odata=nometadata;streaming=true;charset=utf-8

{"odata.error":{"code":"-2146233083, Microsoft.SharePoint.Client.ClientServiceTimeoutException","message":{"lang":"en-US","value":"The operation has timed out."}}}
--batchresponse_202f2ea9-5361-48b3-b92d-c4ead9c25a5a
Content-Type: application/http
Content-Transfer-Encoding: binary
```

**Comment by sympmarc** [comment]
Update on this from my end...

Breaking the update batches up into 100 request chunks seems to be reliably fixing my issues. One suggestion for the [batching capability](https://github.com/SharePoint/PnP-JS-Core/wiki/Batching) would be to add a parameter so that a large batch could be broken into chunks of size x. We'd still get the same results back, but then could mitigate any timeout issues we run into. Maybe something like:

```typescript
let batch = pnp.sp.createBatch(100);
```
where 100 is the number of requests per batch.

The splitting above would also be useful if we need to make more than 100 GET requests, as there seems to be a 100 item limit for that type of batch. (If I'm wrong about this, I'd be happy to be proven such and informed how to do it.) And yes, using batching for GET requests does have some use cases; I could explain mine if it would be helpful.

I've "borrowed" the _parseResponse code from [batch.ts](https://github.com/SharePoint/PnP-JS-Core/blob/06f8ffa9e16005854276e64bcbc39c38f3c1f159/src/sharepoint/batch.ts) and embedded it in my solution. This allows me to keep my existing code rather than converting to use the library entirely. It might be useful to make the parsing capability public so that those of us who build our own batches could still use just the parser. It works very well in the cases where I've used it.

**Comment by russgove** [comment]
I think it would be great if the library could offer some assistance with
batching.
Could be set up like mark said pnp.sp.createBatch(100) or executeBatch(100).

if the dev wanted to control his own batching, then he would just leave off
the batchsize parameter and its up to him to handle it.

**Comment by russgove** [comment]
Any Updates on this?

**Comment by patrick-rodgers** [comment]
I have moved the parse method to a public static on ODataBatch. The below will work after the next release:
```TypeScript
import { ODataBatch } from "sp-pnp-js";

// this should be set with the string body returned from a batch call done outside the library
const batchResponseBody: string = "";

ODataBatch.ParseResponse(batchResponseBody).then((responses:Response[]) => {

    // do stuff with the responses
});
```

As for the batch size parameter, I am unclear on how this would work? If we create a batch using:

```TypeScript
const batch = pnp.sp.createBatch(100);
```
What happens each time 100 items get added? Does the batch auto fire with feedback given to the code? Or would they wait in memory until you called execute and then x batches would all fire (in parallel or sequence)? And how would you handle/process errors? If I have 6 batches of size 100 and execute them all in sequence and one of them has an error what should happen - would we report an error or somehow have to map a series of promises representing each batch?

I am leaning towards this being a sample or placed in the docs as a pattern so that folks who need to process large batches have a reference. It seems like a lot of complication to try and add to the library.

Welcome of course other's thoughts on this topic.

**Comment by sympmarc** [comment]
@patrick-rodgers I think that batching is confusing at best, so at least having some good examples in the docs would be a start.

What's very unclear to me about this is whether there are any "better practices", especially when it comes to SharePoint in particular. I went to 100 requests per batch and it seemed to improve things, but why? What's the ideal batch size? Are there specific limits here imposed by SharePoint's APIs?

In my case, I set up a loop that parses the requests in to chunks of 100...
```javascript
var p = [];
var batchSize = 100;
var batches = Math.ceil(eventIds.length / batchSize);
for (var i = 0; i < batches; i++) {
    var thisBatchItems = eventIds.slice(i * batchSize, ((i * batchSize) + batchSize));
    p.push(IRService.getevents(thisBatchItems));
}
```
...which calls my service function...
```javascript
self.getevents = function (eventIds) {

    var deferred = $q.defer();

    var batchGuid = appService.generateUUID();
    var batchRequest = appService.createBatchRequest(_spPageContextInfo.webAbsoluteUrl,
        self.riskEventsList, batchGuid, eventIds, "Get");

    var request = {
        url: _spPageContextInfo.webAbsoluteUrl + "/_api/$batch",
        method: 'POST',
        headers: {
            "Content-Type": 'multipart/mixed; boundary="batch_' + batchGuid + '"',
            "Accept": "application/json;odata=nometadata",
            "X-RequestDigest": appService.requestDigest
        },
        data: batchRequest
    };

    var success = function (response) {

        var out = self.parseBatch(response.data);

        deferred.resolve(out);
    };

    var error = function (response) {
        deferred.reject(response);
    };

    $http(request).then(success, error);

    return deferred.promise;

};
```
...and concatenates the results.
```javascript
for (var i = 0; i < batches; i++) {
  $scope.events = $scope.events.concat(responses[i]);
}
```

Note that I am using altered versions of AC's functions `generateUUID`, `createBatchRequest` and your `parseBatch`.

Yes, there are some holes here. Error handling is "lite", at best. In my years writing SOAP and REST calls, I've found that well-formed requests simply work - unless the server goes down. At that point, we're usually SOL, anyway. This experience with batching has shown me that requests can indeed fail due to server timeouts, thus my questions above.

**Comment by koltyakov** [comment]
Going to close this. Feel free to reopen.
Also added the example in a [wiki section](https://github.com/SharePoint/PnP-JS-Core/wiki/Batching#large-batch-processing).
